src/Engine/Component/ParticleComponents/ParticleEffect.cpp

```diff
@@ -0,0 +1,393 @@
+    m_activeParticleOrientationIDs(getActiveParticleOrientationIDs()),
+    m_activeMap(getActiveMap()),
+    m_nextActivation(getNextActivation()),
+    m_attenuationRate(1 / 60.0f),
+    m_nextAttenuation(m_attenuationRate),
+    m_life(0.0f)
+{
+}
+
+ParticleEffect::ParticleEffect(EffectParams *effectParams, const glm::vec3 & offset, const glm::vec3 & direction) :
+    m_effectParams(effectParams),
+    m_anchor(glm::vec3(0)),
+    m_offset(offset),
+    m_direction(direction),
+    m_velocity(glm::vec3(0.0f, 0.0f, 0.0f)),
+    m_particles(generateParticles()),
+    m_activeParticlePositions(getActiveParticlePositions()),
+    m_activeParticleOrientationIDs(getActiveParticleOrientationIDs()),
+    m_activeMap(getActiveMap()),
+    m_nextActivation(getNextActivation()),
+    m_attenuationRate(1 / 60.0f),
+    m_nextAttenuation(m_attenuationRate),
+    m_life(0.0f)
+{
+}
+
+ParticleEffect::ParticleEffect(EffectParams *effectParams, const glm::vec3 & offset, const glm::vec3 & direction, 
+    const glm::vec3 & velocity) :
+    m_effectParams(effectParams),
+    m_anchor(glm::vec3(0)),
+    m_offset(offset),
+    m_direction(direction),
+    m_velocity(velocity),
+    m_particles(generateParticles()),
+    m_activeParticlePositions(getActiveParticlePositions()),
+    m_activeParticleOrientationIDs(getActiveParticleOrientationIDs()),
+    m_activeMap(getActiveMap()),
+    m_nextActivation(getNextActivation()),
+    m_attenuationRate(1 / 60.0f),
+    m_nextAttenuation(m_attenuationRate),
+    m_life(0.0f)
+{
+}
+
+void ParticleEffect::update(float dt, glm::vec3 anchor, glm::vec3 direction) {
+    m_anchor = anchor;
+    m_direction = direction;
+    
+    update(dt);
+}
+
+void ParticleEffect::update(float dt, glm::vec3 anchor) {
+    m_anchor = anchor;
+    update(dt);
+}
+
+void ParticleEffect::update(float dt) {
+    m_life += dt;
+
+    if (m_life < m_effectParams->effectDuration) {
+        updateActiveParticles(dt);
+        int attenuations = getAttenuations(dt);
+        float factor = 1.0f;
+        if (attenuations != 0) {
+            factor= pow(m_effectParams->attenuation, attenuations);
+        }
+        for (int i = 0; i < (int)m_activeParticlePositions->size(); i++) {
+            Particle *p = m_particles[m_activeMap[i]];
+            updatePosition(p, dt, factor);
+            m_activeParticlePositions->at(i) = p->position;
+            p->life += dt;
+
+        }
+    }
+    else if (m_effectParams->loop) {
+        if (m_effectParams->rate == 0) {
+            m_life = 0.0f;
+            m_particles = generateParticles();
+            m_activeParticlePositions = getActiveParticlePositions();
+            m_activeParticleOrientationIDs = getActiveParticleOrientationIDs();
+            m_activeMap = getActiveMap();
+            m_nextActivation = getNextActivation();
+            m_nextAttenuation = 0.0f;
+        }
+        else {
+            m_effectParams->effectDuration += m_life;
+        }
+        update(dt);
```

Comment:
Recursive update? Why?

Reply:
This is so when a particle effect is over and it is loops, it starts from the get go rather than starting one update late.

Reply:
??? how is this not infinite recursion? A component should never call its own update method